compiler/rustc_hir_analysis/src/errors.rs

```diff
@@ -1378,54 +1379,47 @@ pub(crate) struct CrossCrateTraitsDefined {
     pub traits: String,
 }
 
-// FIXME(fmease): Deduplicate:
-
-#[derive(Diagnostic)]
-#[diag(hir_analysis_ty_param_first_local, code = E0210)]
-#[note]
-pub(crate) struct TyParamFirstLocal<'tcx> {
-    #[primary_span]
-    #[label]
-    pub span: Span,
-    #[note(hir_analysis_case_note)]
-    pub note: (),
-    pub param: Symbol,
-    pub local_type: Ty<'tcx>,
-}
-
-#[derive(LintDiagnostic)]
-#[diag(hir_analysis_ty_param_first_local, code = E0210)]
-#[note]
-pub(crate) struct TyParamFirstLocalLint<'tcx> {
-    #[label]
-    pub span: Span,
-    #[note(hir_analysis_case_note)]
-    pub note: (),
-    pub param: Symbol,
-    pub local_type: Ty<'tcx>,
+pub(crate) struct UncoveredTy<'tcx, T> {
+    pub span: Span,
+    pub kind: (&'static str, &'static str),
+    pub ty: T,
+    pub local_ty: Option<Ty<'tcx>>,
+}
+
+impl<'tcx, T: IntoDiagArg> UncoveredTy<'tcx, T> {
+    fn decorate<G: EmissionGuarantee>(self, diag: &mut Diag<'_, G>) {
+        diag.primary_message(match self.local_ty {
+            Some(_) => fluent::hir_analysis_uncovered_ty_before_first_local,
+            None => fluent::hir_analysis_uncovered_ty,
+        });
+        diag.code(E0210);
+        diag.span_label(self.span, fluent::hir_analysis_label);
+        diag.note(fluent::hir_analysis_note);
+        diag.note(match self.local_ty {
+            Some(_) => fluent::hir_analysis_case_note,
+            None => fluent::hir_analysis_only_note,
+        });
+        diag.arg("article", self.kind.0);
+        diag.arg("kind", self.kind.1);
+        diag.arg("ty", self.ty);
+        if let Some(local_ty) = self.local_ty {
+            diag.arg("local_ty", local_ty);
+        }
+    }
```

Comment:
if you've changing this anyways, do you mind changing this back into non-translateable diagnostics? :thinking: it feels somewhat hard to review otherwise cc #132181